### ISSUE TITLE
Handle 'Unexpected "space" found.' Error better

### ISSUE THREAD

**Issue by Arcanemagus** [stylelint/stylelint#1153]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

There is a bug in `stylelint` when parsing some files where it throws an `Unexpected "space" found.` error.

> Which rule, if any, is this issue related to?

Apparently this is a bug in the following rules:

* `selector-type-case`
* `selector-pseudo-element-case`
* `selector-pseudo-class-case`

> What CSS is needed to reproduce this issue?

```scss
.foo {
  -moz-osx-font-smoothing: grayscale
  .bar & {}
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-pseudo-element-case": "lower",
    "selector-pseudo-class-case": "lower",
    "selector-type-case": "lower"
  }
}
```

Note: Any _one_ of the above rules is enough to trigger this.

> Which version of stylelint are you using?

v6.2.2

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI and Node API (`linter-stylelint`)

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Discovered in SCSS, but shows if the file is parsed as CSS.

> What did you expect to happen?

`stylelint` to catch this error and return it in the normal manner messages are returned, instead of `throw`ing it on up to where `stylelint` was being called.

> What actually happened (e.g. what warnings or errors you are getting)?

The following stack trace:

```raw
Error: Unexpected "space" found.
    at new error (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\processor.js:27:23)
    at Parser.error (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\parser.js:212:15)
    at Parser.pseudo (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\parser.js:320:18)
    at Parser.parse (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\parser.js:502:22)
    at Parser.loop (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\parser.js:470:18)
    at new Parser (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\parser.js:89:21)
    at Processor.process (C:\Temp\github\linter-stylelint_GH166\node_modules\postcss-selector-parser\dist\processor.js:24:21)
    at C:\Temp\github\linter-stylelint_GH166\node_modules\stylelint\dist\rules\selector-pseudo-element-case\index.js:48:10
    at C:\Temp\github\linter-stylelint_GH166\node_modules\postcss\lib\container.js:110:28
    at C:\Temp\github\linter-stylelint_GH166\node_modules\postcss\lib\container.js:73:26
```

Note: Originally filed as part of https://github.com/AtomLinter/linter-stylelint/issues/166.

**Comment by evilebottnawi** [comment]
@Arcanemagus incorrect scss syntax
```scss
.foo {
    -moz-osx-font-smoothing: grayscale; // here need `;`
    .bar & {}
}
```

**Comment by davidtheclark** [comment]
@evilebottnawi I think the issue here is that the user wants a better UX about this error, instead of having it popup in a weird way in Atom. What we would have to do is catch the error that postcss-selector throws and then output it is as a warning; Atom linter would then display it in a less obstructive way.

**Comment by evilebottnawi** [comment]
@davidtheclark you right

**Comment by ben-eb** [comment]
Agreed that it should be handled here rather than in selector-parser - watch out for false positives however (someone accidentally writes `a: before` instead of `a:before`).

**Comment by Arcanemagus** [comment]
@davidtheclark Is absolutely correct here, ideally this error message would be caught and returned as a "normal" error, instead of being thrown up the chain causing it to be caught as an error in stylelint that has no position (we treat those as _major_ issues with trying to run `stylelint` and alert the user to them, unless they have a line associated with them).

I've updated the "What did you expect to happen?" to better reflect that.

**Comment by Arcanemagus** [comment]
https://github.com/AtomLinter/linter-stylelint/issues/183 isn't the same root cause of the dialog popping up, but the solution to this will probably also solve that as the error would be shown normally to the user.

**Comment by Arcanemagus** [comment]
@evilebottnawi I'm curious what this is blocked on?

**Comment by jeddy3** [comment]
I think that label was related waiting on the question issue in `postcss-selector-parser`. I've updated the label accordingly :)

**Comment by smlombardi** [comment]
This is driving me nuts.  I am using the Atom linter-stylelint on my scss files.  Every time I type a space after a semi-colon, I get a full-blown error dialog:

![screen shot 2016-05-09 at 9 40 35 am](https://cloud.githubusercontent.com/assets/835148/15114941/680f0bb0-15ca-11e6-99be-4b7478331570.jpg)

The authors of this linter say this is because stylelint does not return a cursor position in these cases, so they cannot display the smaller inline error tooltip.

Hopefully this gets fixed.  IMO we don't need an "unexpected space" while we're in the process of typing a perfectly legal space anyway,

**Comment by davidtheclark** [comment]
Feel free to pitch in. Otherwise, you can always change your linter plugin to run on save instead of on every change.

**Comment by kevinkorngut** [comment]
I'm running into this as well via the command line; however, my config looks like this: 

```
{
  extends: [
    'stylelint-config-standard'
  ],
  rules: {
    "at-rule-no-vendor-prefix": true,
    "media-feature-name-no-vendor-prefix": true,
    "property-no-vendor-prefix": true,
    "selector-no-vendor-prefix": true,
    "value-no-vendor-prefix": true
  }
}
```
and the css:
```
.hello-world {
  max-width: 95vw;

  & div {
    display: block;
  }
}
```

Now, if I remove the semi-colon on the ```max-width``` line I receive the "unexpected space" error; however, if I remove it from the ```display``` line I get an error from the ```declaration-block-trailing-semicolon``` rule which is what I would have expected.

**Comment by davidtheclark** [comment]
Yeah, we're aware of the error. The probably reason is written about above.

**Comment by evilebottnawi** [comment]
@davidtheclark what do you think about solution this issue? Create wrapper function in utils with `try {} catch {}` and import this util instead directly `import selectorParser from "postcss-selector-parser"`?

**Comment by davidtheclark** [comment]
Yeah, I think so. What would be the best warning message to deliver? Just something like "Cannot parse selector"?

**Comment by evilebottnawi** [comment]
@davidtheclark i think we have nothing to output, because when fast typing such errors can be many. While on the other hand, it is not entirely correct.

**Comment by smlombardi** [comment]
@davidtheclark I'd pitch in if I understood any of this.  Sorry, I'm just a css dude.  ;-)

If the "lint on save" could be done on a particular linter level I'd do that, but AFAIK it's a global setting in Atom for all linters, and for all my other linters I like the linting as I type.

**Comment by smlombardi** [comment]
@evilebottnawi Is there a concept here like "debounce", where you can delay the error for an amount of time or amount of keystrokes?

**Comment by davidtheclark** [comment]
I will take a shot at this today.

**Comment by Arcanemagus** [comment]
@smlombardi `stylelint` has no idea what is going on in the editor, it's just being called with a block of text and asked to tell `linter-stylelint` what is wrong with it (which then translates that and sends it on to `linter`).

You might want to bump up the debounce timer inside `linter` from the default, I personally have mine at 1500 ms.

**Comment by smlombardi** [comment]
@Arcanemagus Done.  Great tip, thanks.
